perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -1084,200 +1087,214 @@ protected boolean hasCandidateExistingMember(Candidate candidate, RichMember ric
-				startTime = System.currentTimeMillis();
+				try {
+					// Set the start time, so we can check the timeout of the thread
+					startTime = System.currentTimeMillis();
+
+					log.debug("Synchronization thread started synchronization for group {}.", group);
+
+					//synchronize Group and get information about skipped Members
+					List<String> skippedMembers = perunBl.getGroupsManagerBl().synchronizeGroup(sess, group);
 
-				//synchronize Group and get information about skipped Members
-				List<String> skippedMembers = perunBl.getGroupsManagerBl().synchronizeGroup(sess, group);
+					if (!skippedMembers.isEmpty()) {
+						skippedMembersMessage = "These members from extSource were skipped: { ";
 
-				if(!skippedMembers.isEmpty()) {
-					skippedMembersMessage = "These members from extSource were skipped: { ";
+						for (String skippedMember : skippedMembers) {
+							if (skippedMember == null) continue;
 
-					for(String skippedMember: skippedMembers) {
-						if(skippedMember == null) continue;
+							skippedMembersMessage += skippedMember + ", ";
+						}
+						skippedMembersMessage += " }";
+						exceptionMessage = skippedMembersMessage;
+					}
 
-						skippedMembersMessage+= skippedMember + ", ";
+					log.debug("Synchronization thread for group {} has finished in {} ms.", group, System.currentTimeMillis() - startTime);
+				} catch (WrongAttributeValueException | WrongReferenceAttributeValueException | InternalErrorException |
+						WrongAttributeAssignmentException | MemberAlreadyRemovedException | GroupNotExistsException |
+						GroupOperationsException | AttributeNotExistsException | ExtSourceNotExistsException e) {
+					failedDueToException = true;
+					exceptionMessage = "Cannot synchronize group ";
+					log.error(exceptionMessage + group, e);
+					exceptionMessage += "due to exception: " + e.getName() + " => " + e.getMessage();
+				} catch (Exception e) {
+					//If some other exception has been thrown, log it and throw again
+					failedDueToException = true;
+					exceptionMessage = "Cannot synchronize group ";
+					log.error(exceptionMessage + group, e);
+					exceptionMessage += "due to unexpected exception: " + e.getClass().getName() + " => " + e.getMessage();
+				} finally {
+					//Save information about group synchronization, this method run in new transaction
+					try {
+						perunBl.getGroupsManagerBl().saveInformationAboutGroupSynchronization(sess, group, failedDueToException, exceptionMessage);
+					} catch (Exception ex) {
+						log.error("When synchronization group " + group + ", exception was thrown.", ex);
+						log.info("Info about exception from synchronization: " + skippedMembersMessage);
+					}
+					//Remove job from running jobs
+					try {
+						poolOfGroupsToBeSynchronized.removeJob(group);
+					} catch (Exception ex) {
+						log.error("Can't remove running job from pool of running jobs. Job for group " + group + " is stuck!", ex);
```

Comment:
Could you provide more information about this error?
What do you mean by "job is stuck"? Any recommended actions, possible causes, ...